File: 2022_11_03_leetcode_in_fsharp/LeetCode/P0118_PascalsTriangle.fs
﻿// url: https://leetcode.com/problems/pascals-triangle/
// tags: unfold, pairwise, take
// examples: 2 -> [ [ 1 ], [ 1, 1 ] ]

module LeetCode.P0118_PascalsTriangle

open Utils

let pascalsTriangle rowNumber =
    [ 1 ]
    |> Seq.unfold (fun prev ->
        let lst =
            List.concat [ [ 1 ]
                          prev |> List.pairwise |> List.map (fun (x, y) -> x + y)
                          [ 1 ] ]
        Some(lst, lst))
    |> Seq.append [ [ 1 ] ]
    |> Seq.take rowNumber
    |> Seq.toArray

pascalsTriangle 0 === [||]
pascalsTriangle 1 === [| [ 1 ] |]
pascalsTriangle 2 === [| [ 1 ]; [ 1; 1 ] |]
pascalsTriangle 3 === [| [ 1 ]; [ 1; 1 ]; [ 1; 2; 1 ] |]
pascalsTriangle 4 === [| [ 1 ]; [ 1; 1 ]; [ 1; 2; 1 ]; [ 1; 3; 3; 1 ] |]
pascalsTriangle 5 === [| [ 1 ]; [ 1; 1 ]; [ 1; 2; 1 ]; [ 1; 3; 3; 1 ]; [ 1; 4; 6; 4; 1 ] |]

File: 2022_11_19_leetcode_in_csharp/LeetCode/P0118_PascalsTriangle.cs
// https://leetcode.com/problems/pascals-triangle/

namespace LeetCode;

using static System.Linq.Enumerable;

class P0118_PascalsTriangle
{
    static int[][] PascalsTriangle(int rowNumber)
    {
        var one = new[] { 1 };

        return new[] { one }
            .Expand(prev => new[] { one.Concat(prev.Pairwise().Select(t => t.Item1 + t.Item2)).Concat(one).ToArray() })
            .Take(rowNumber)
            .ToArray();
    }

    public static void Run()
    {
        var inputs = new[] { 0, 1, 2, 3, 4, 5 };

        foreach (var input in inputs)
        {
            var output = PascalsTriangle(input);
            Console.WriteLine($" {input} -> {output.Select(row => row.JoinToString()).JoinToString(" ")}");
        }
    }
}

